macros/src/lib.rs

```diff
@@ -0,0 +1,74 @@
+use proc_macro::TokenStream;
+
+#[proc_macro_attribute]
+pub fn machine(_attrs: TokenStream, fun: TokenStream) -> TokenStream {
+    let f = syn::parse_macro_input!(fun as syn::ItemFn);
+
+    assert!(
+        f.attrs.is_empty(),
+        "netsim_embed::machine cannot have any attribute"
+    );
+    assert!(
+        f.sig.constness.is_none(),
+        "netsim_embed::machine cannot be const"
+    );
+    assert!(
+        f.sig.asyncness.is_none(),
+        "netsim_embed::machine cannot be async"
+    );
+    assert!(
+        f.sig.unsafety.is_none(),
+        "netsim_embed::machine cannot be unsafe"
+    );
+    assert!(
+        f.sig.abi.is_none(),
+        "netsim_embed::machine cannot have an abi defined"
+    );
+    assert!(
+        f.sig.generics.params.is_empty(),
+        "netsim_embed::machine cannot be generic"
+    );
+    assert!(
+        f.sig.inputs.len() == 1,
+        "netsim_embed::machine must take exactly one argument"
+    );
+    assert!(
+        f.sig.variadic.is_none(),
+        "netsim_embed::machine cannot be variadic"
+    );
+    assert!(
+        matches!(f.sig.output, syn::ReturnType::Default),
+        "netsim_embed::machine must not declare a return type"
+    );
+
+    let input = match &f.sig.inputs.first().unwrap() {
+        syn::FnArg::Typed(input) => input,
+        _ => panic!("netsim_embed::machine must be a freestanding function"),
+    };
+    assert!(
+        input.attrs.is_empty(),
+        "netsim_embed::machine's only argument must not have any attributes attached"
+    );
+
+    let f_vis = f.vis;
+    let f_ident = f.sig.ident;
+    let input_ty = &input.ty;
+    let id: u128 = rand::random();
```

Comment:
Do I understand correctly that this randomness does not affect when/if the tests will be recompiled? Can this be an issue with rust-analyzer or similar tools? This has a bit of a quantum mechanic vibe to it: just looking at this thing changes it ;-)

Reply:
You understand correctly! I just tried running the tests three times in a row to be absolutely sure; seeing how unlikely it is to get two u128 the same I'm pretty sure I'm correct :)

In practice, this is just an impurity in the build; similar to when proc macros use HashMap's or similar. It will result in different binaries, but given the dependency on a random number is not actually tracked, it will not trigger a rebuild.

This is definitely something I would do my best to avoid if it were in a binary expected to be deployed to production (because, as a nixos user, I'm striving for build purity). But netsim-embed and especially this function are designed to only ever be used for tests, so I don't think it's a big deal if someone someday decides to use it in production and then complains that building twice bitwise-doesn't give the same binary – there are most likely many other sources of impurity that would be more important to tackle than this one, and this one actually offers us more type-safety :)